### PR TITLE
Resolve #206: Add ability to download files from the browser.

### LINKED ISSUE #206: Enable downloading of files into the Downloads folder
Enable the UI for downloading of files into the Downloads folder by reverting changes made within #205

### PR DESCRIPTION
**Pull request by iccub**
All of file download logic and Downloads panel is taken from Firefox.
What was implemented by me:
- download prompt is new
- new menu entry for downloads panel

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

**Comment by jumde** [comment]
Running into a crash

```
(lldb) bt 5
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x000000021a8990dc libsystem_kernel.dylib`__pthread_kill + 8
    frame #1: 0x000000021a912094 libsystem_pthread.dylib`pthread_kill$VARIANT$mp + 380
    frame #2: 0x000000021a7f2ea8 libsystem_c.dylib`abort + 140
    frame #3: 0x0000000219ebf788 libc++abi.dylib`abort_message + 132
    frame #4: 0x0000000219ebf934 libc++abi.dylib`default_terminate_handler() + 308
```

STR:
1. Navigate to http://cyberside.net.ee/ripping/
2. Click on 1click-bluray-copy_v1.0.0.6beta.exe

**Comment by jumde** [comment]
- [ ] Downloads in the private-tab are visible in the download panel in the normal tab. I think its okay to keep the files, but we should remove the reference from the normal ( mimics the behavior on desktop )

**Comment by jumde** [comment]
- [ ] Clear Private Data does not clear downloads.

**Comment by jhreis** [comment]
I am blocking this beyond security review. Please DO NOT MERGE, until talking with me. There are some timing mechanics that need to be worked out.

File: Client/Info.plist
Comment (jhreis) [inline Client/Info.plist]: This needs to be a bool `<true/>`, will probably need to edit the plaintext, as I see Xcode has it stuck as a `string` type.

File: Client/Frontend/Browser/DownloadQueue.swift
Comment (jhreis) [inline Client/Frontend/Browser/DownloadQueue.swift:164]: This seems not ideal. We should either:
- Pass in the delegate to the constructor
- Use Deferred, and call the delegate on `fill`, then if this fills the request, when the delegate is attached, it is automatically fired

File: BraveShared/BraveStrings.swift
Comment (jhreis) [inline BraveShared/BraveStrings.swift]: I think this `comment` could be a bit more clear, it is for translators, so something more like:
> Text to select sharing something (example: image, video, URL)

File: BraveShared/BraveStrings.swift
Comment (jhreis) [inline BraveShared/BraveStrings.swift]: Same as above, maybe:
> Text to choose for downloading a file (example: saving an image to phone)

File: BraveShared/BraveStrings.swift
Comment (jhreis) [inline BraveShared/BraveStrings.swift]: comment maybe:
> Title for when a user has nothing downloaded onto their device, and the list is empty.

File: Shared/DateGroupedTableData.swift
Comment (jhreis) [inline Shared/DateGroupedTableData.swift:1]: Is this sourced from somewhere?

**Review by jhreis — CHANGES_REQUESTED** [review]

**Comment by jhreis** [comment]
Talked with @jumde, we can add a new check list item in Clear Private Data specifically for "downloads".

File: Shared/DateGroupedTableData.swift
Comment (iccub) [reply Shared/DateGroupedTableData.swift:1]: It's copy pasted from https://github.com/mozilla-mobile/firefox-ios/blob/master/Shared/DateGroupedTableData.swift

File: Client/Frontend/Browser/BrowserViewController.swift
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController.swift]: Remove unnecessary parens.

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:238]: What was the reason for relocating this logic above the url code below?

--
(if this was moved over from FF, ignore the comment below)
I know this existed before, but since this is just inverted of part of the logic above maybe:
```
        let forceDownload = webView == pendingDownloadWebView
        let canShowInWebView = navigationResponse.canShowMIMEType && !forceDownload
```

**Review by jhreis — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (iccub) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:238]: moved from FF

File: Client/Frontend/Browser/DownloadQueue.swift
Comment (iccub) [reply Client/Frontend/Browser/DownloadQueue.swift:164]: This is what FF does and only for Blob files, we don't even support them yet.
We will implement it in a new ticket https://github.com/brave/brave-ios/issues/1424

Comment (jhreis) [reply Client/Frontend/Browser/DownloadQueue.swift:164]: Perfect

**Review by jhreis — APPROVED** [review]

**Comment by jhreis** [comment]
We are still blocked on:
- [ ] security review
- [x] new copy for private browsing disclaimer text
cc'ing @tomlowenthal @jumde @anthonypkeane

**Comment by jumde** [comment]
Updated text: 

```
Private Tabs aren’t saved in Brave, but they don’t make you anonymous online. Sites you visit in a private tab won’t show up in your history and their cookies always vanish when you close them — there won’t be any trace of them left in Brave. However, downloads will be saved.

Your mobile carrier (or the owner of the WiFi network or VPN you’re connected to) can see which sites you visit and those sites will learn your public IP address, even in Private Tabs.
```

**Review by jumde — APPROVED** [review]

**Comment by iccub** [comment]
Updated text

Zrzut ekranu 2019-08-27 o 00 17 23

**Pull request merged**
